src/util/vec2d.rs

```diff
@@ -0,0 +1,137 @@
+use std::ops::{Index, IndexMut};
+
+/// A 2 dimensional matrix in row-major order backed by a contiguous `Vec`
+#[derive(Debug)]
+pub struct Vec2D<T> {
+    data: Box<[T]>,
+    cols: usize,
+}
+
+impl<T> Vec2D<T> {
+    /// Initialize a grid of size (`rows`, `cols`) with the given data element.
+    pub fn init(data: T, size: (usize, usize)) -> Vec2D<T>
+    where
+        T: Clone,
+    {
+        let (rows, cols) = size;
+        let len = rows
+            .checked_mul(cols)
+            .unwrap_or_else(|| panic!("{} rows by {} cols exceeds usize::MAX", rows, cols));
+        Vec2D {
+            data: vec![data; len].into_boxed_slice(),
+            cols,
+        }
+    }
+
+    /// Fills the grid with elements by cloning `value`.
+    pub fn fill(&mut self, value: T)
+    where
+        T: Clone,
+    {
+        self.data.fill(value)
+    }
+}
+
+impl<T> Index<usize> for Vec2D<T> {
+    type Output = [T];
+
+    #[inline]
+    fn index(&self, row: usize) -> &Self::Output {
+        let start_row = row * self.cols;
+        &self.data[start_row..start_row + self.cols]
+    }
+}
+
+impl<T> IndexMut<usize> for Vec2D<T> {
+    #[inline]
+    fn index_mut(&mut self, row: usize) -> &mut Self::Output {
+        let start_row = row * self.cols;
+        &mut self.data[start_row..start_row + self.cols]
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn vec2d_init() {
+        let vec2d = Vec2D::init(1, (2, 3));
+        assert_eq!(vec2d[0], [1, 1, 1]);
+        assert_eq!(vec2d[1], [1, 1, 1]);
+    }
+
+    #[test]
+    fn vec2d_fill() {
+        let mut vec2d = Vec2D::init(0, (2, 3));
+        vec2d.fill(7);
+        assert_eq!(vec2d[0], [7, 7, 7]);
+        assert_eq!(vec2d[1], [7, 7, 7]);
+    }
+
+    #[test]
+    fn vec2d_index() {
+        let vec2d = Vec2D {
+            data: vec![0, 1, 2, 3, 4, 5, 6, 7, 8].into_boxed_slice(),
+            cols: 3,
```

Comment:
I wouldn't use the same number of rows and columns in this test, so that it's obvious in this test what are the rows vs. column dimensions.